text/000-sfc-and-template-import-primitives.md

```diff
@@ -0,0 +1,679 @@
+        <OtherComponent />
+      </div>
+    );
+  }
+}
+```
+
+This is a bit nicer than Vue's SFCs, where JavaScript code and template code
+happen to be in the same file but otherwise interact no differently than if
+they were still in separate files. Developers must learn a proprietary Vue
+API for explicitly copying values from JavaScript's scope into the template's
+scope, and component names can be different between where they are imported
+(`OtherComponent`) and where they are invoked (`other-component`):
+
+```html
+<template>
+  <div>
+    <other-component />
+  </div>
+</template>
+
+<script>
+import OtherComponent from './OtherComponent.vue'
+
+export default {
+  components: {
+    OtherComponent
+  }
+}
+</script>
+```
+
+Neither of these approaches is exactly right for Ember. Ideally, we'd find a
+way to "bend the curve" of tradeoffs: maintaining the performance benefits of
+templates, while gaining productivity and learnability by having those
+templates seamlessly participate in JavaScript's scoping rules.
+
+### Unlocking Experimentation
+
+Rather than deciding on the best syntax for single-file components and
+template imports upfront, this RFC proposes new low-level APIs that addons
+can use as compile targets to implement experimental file formats.
+
+For example, imagine a file format that uses a frontmatter-like syntax to
+combine template and JavaScript sections in a single file:
+
+```hbs
+---
+import Component from '@glimmer/component';
+import { inject as service } from '@ember/service';
+import titleize from './helpers/titleize';
+import BlogPost from './components/blog-post';
+
+export default class MyComponent extends Component {
+  @service session;
+}
+---
+
+{{#let this.session.currentUser as |user|}}
+  <BlogPost @title={{titleize @model.title}} @body={{@model.body}} @author={{user}} />
+{{/let}}
+```
+
+Or imagine this syntax, where a component's template is located syntactically
+within the class body:
+
+```js
+import Component from '@glimmer/component';
+import { inject as service } from '@ember/service';
+import titleize from './helpers/titleize';
+import BlogPost from './components/blog-post';
+
+export default class MyComponent extends Component {
+  <template>
+    {{#let this.session.currentUser as |user|}}
+      <BlogPost @title={{titleize @model.title}} @body={{@model.body}} @author={{user}} />
+    {{/let}}
+  </template>
+    
+  @service session;
+}
+```
+
+Or even a variation of the Vue SFC format:
+
+```html
+<template>
+{{#let this.session.currentUser as |user|}}
+  <BlogPost @title={{titleize @model.title}} @body={{@model.body}} @author={{user}} />
+{{/let}}
+</template>
+
+<script type="module">
+import Component from '@glimmer/component';
+import { inject as service } from '@ember/service';
+import titleize from './helpers/titleize';
+import BlogPost from './components/blog-post';
+
+export default class MyComponent extends Component {
+  @service session;
+}
+</script>
+```
```

Comment:
I think it might be worth adding an example which shows how a syntax could facilitate both inline and separate files. Although this RFC is only intended to facilitate experimentation, separation of concerns has long been a tenant of Ember's and is an important use case for many users. An example showing how such an API could unlock easy inline templates, while not excluding dedicated template files, would probably go a long way to making people feel more comfortable about exploring this direction.

#### For example (this is completely contrived, but hopefully it gets the point across):

`my-component/component.js`
```js
import Component from '@glimmer/component';
import { inject as service } from '@ember/service';
import titleize from './helpers/titleize';
import BlogPost from './components/blog-post';

export default class MyComponent extends Component {
  @service session;

  template() {
    return `
      {{#let this.session.currentUser as |user|}}
        <BlogPost @title={{titleize @model.title}} @body={{@model.body}} @author={{user}} />
      {{/let}}
    `;
  };
}
```

#### Could also be written as:

`my-component/template.hbs`
```hbs
{{#let this.session.currentUser as |user|}}
  <BlogPost @title={{titleize @model.title}} @body={{@model.body}} @author={{user}} />
{{/let}}
```

`my-component/component.js`
```js
import Component from '@glimmer/component';
import { inject as service } from '@ember/service';
import titleize from './helpers/titleize';
import BlogPost from './components/blog-post';
import Template from './template.hbs';

export default class MyComponent extends Component {
  @service session;
  template() { return Template };
}
```

I'm aware there's some problems with the second example, such as `BlogPost` and `titleize` not being utilized within the component directly, problems which this RFC aims to help address by opening up experimentation. However, I think it would be helpful to demonstrate that such experimentation doesn't necessarily lead us to a place where we can't support both inline and separate file templates.

Reply:
FWIW I believe that these would be the corresponding examples for the variations. I do agree having them be available in this RFC would be nice, since this is also about how to move forward with importing in non-SFC components in general

### Frontmatter

`my-component/template.hbs`

```hbs
---
import titleize from './helpers/titleize';
import BlogPost from './components/blog-post';
---

{{#let this.session.currentUser as |user|}}
  <BlogPost @title={{titleize @model.title}} @body={{@model.body}} @author={{user}} />
{{/let}}
```

`my-component/component.js`
```js
import Component from '@glimmer/component';
import { inject as service } from '@ember/service';

export default class MyComponent extends Component {
  @service session;
}
```

### Inline

`my-component/template.js`
```js
import titleize from './helpers/titleize';
import BlogPost from './components/blog-post';

<template>
  {{#let this.session.currentUser as |user|}}
    <BlogPost @title={{titleize @model.title}} @body={{@model.body}} @author={{user}} />
  {{/let}}
</template>
```

`my-component/component.js`
```js
import Component from '@glimmer/component';
import { inject as service } from '@ember/service';

export default class MyComponent extends Component {
  @service session;
}
```

### Vue Style

`my-component/template.hbs`
```hbs
<script type="module">
  import titleize from './helpers/titleize';
  import BlogPost from './components/blog-post';
</script>
<template>
  {{#let this.session.currentUser as |user|}}
    <BlogPost @title={{titleize @model.title}} @body={{@model.body}} @author={{user}} />
  {{/let}}
</template>
```

`my-component/component.js`
```js
import Component from '@glimmer/component';
import { inject as service } from '@ember/service';

export default class MyComponent extends Component {
  @service session;
}
```

Reply:
Thanks for the better examples @pzuraq!